sql: more routine dependency fixes

DrewKimball opened the pull request:
#### sql: track SET columns in routine dependencies

Previously, we did not track dependencies on target columns in the SET clause
of an UPDATE statement while creating a routine. This meant that it was
possible to break a routine by dropping such a column if it wasn't referenced
elsewhere in the routine. This commit fixes the bug by adding target columns
that are explicitly referenced by an UPDATE statement to the routine
dependencies. The new behavior is gated behind the session variable
`prevent_update_set_column_drop`.

Fixes #158898

Release note (bug fix): Fixed a bug that allowed a column to be dropped from
its table despite being referenced by a routine. The bug could happen when the
column was only referenced as a target column in the SET clause of an UPDATE
statement within the routine. This fix only applies to newly-created routines.
In versions prior to v26.1, the fix must be enabled by setting the session
variable `prevent_update_set_column_drop`.

#### sql/logictest: rearrange trigger logic test

This commit pulls out some re-arranging of the `triggers` logic test
to avoid a confusing diff in the following commmit.

Epic: None
Release note: None

#### sql: avoid unnecessary column dependencies in routines

Previously, there were several cases where routines would add unnecessary
column dependencies:
* Columns referenced by a computed column expression would be added to
  the list of columns referenced by the routine.
* BEFORE triggers would confuse the logic used to distinguish explicit
  from implicit INSERT columns, because triggers change the column IDs
  to be inserted. This could cause the routine to add implicitly referenced
  columns to its dependencies.
* Dependencies from statements within a trigger would transitively apply
  to a routine that invoked trigger on a mutation.

This commit prevents newly-created routines from adding unnecessary
column dependencies in those cases, gated behind the session setting
`use_improved_routine_deps_triggers_and_computed_cols`.

Fixes #158154

Release note (bug fix): Fixed a bug that caused routines to prevent
dropping more columns than necessary, most notably columns referenced
by computed column expressions. The fix is gated behind the session
setting `use_improved_routine_deps_triggers_and_computed_cols`, which
is off by default prior to v26.1.

DrewKimball committed:
sql: track SET columns in routine dependencies

Previously, we did not track dependencies on target columns in the SET clause
of an UPDATE statement while creating a routine. This meant that it was
possible to break a routine by dropping such a column if it wasn't referenced
elsewhere in the routine. This commit fixes the bug by adding target columns
that are explicitly referenced by an UPDATE statement to the routine
dependencies. The new behavior is gated behind the session variable
`prevent_update_set_column_drop`.

Fixes #158898

Release note (bug fix): Fixed a bug that allowed a column to be dropped from
its table despite being referenced by a routine. The bug could happen when the
column was only referenced as a target column in the SET clause of an UPDATE
statement within the routine. This fix only applies to newly-created routines.
In versions prior to v26.1, the fix must be enabled by setting the session
variable `prevent_update_set_column_drop`.

DrewKimball committed:
sql/logictest: rearrange trigger logic test

This commit pulls out some re-arranging of the `triggers` logic test
to avoid a confusing diff in the following commmit.

Epic: None
Release note: None

DrewKimball committed:
sql: avoid unnecessary column dependencies in routines

Previously, there were several cases where routines would add unnecessary
column dependencies:
* Columns referenced by a computed column expression would be added to
  the list of columns referenced by the routine.
* BEFORE triggers would confuse the logic used to distinguish explicit
  from implicit INSERT columns, because triggers change the column IDs
  to be inserted. This could cause the routine to add implicitly referenced
  columns to its dependencies.
* Dependencies from statements within a trigger would transitively apply
  to a routine that invoked trigger on a mutation.

This commit prevents newly-created routines from adding unnecessary
column dependencies in those cases, gated behind the session setting
`use_improved_routine_deps_triggers_and_computed_cols`.

Fixes #158154

Release note (bug fix): Fixed a bug that caused routines to prevent
dropping more columns than necessary, most notably columns referenced
by computed column expressions. The fix is gated behind the session
setting `use_improved_routine_deps_triggers_and_computed_cols`, which
is off by default prior to v26.1.

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Very nice!

Unfortunately, I think the safe thing to do is make a new session setting given that `use_improved_routine_dependency_tracking` is already on by default in 25.3.

> @mgartner reviewed 3 of 3 files at r1, 1 of 1 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-:-Ofl7k2SEAvFZakFEx7R:b-prnzh4)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball and @ZhouXing19)

___
*[`pkg/sql/logictest/testdata/logic_test/udf_update` line 382 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-Ofl5Dh2-LGyABy4ImRS:-Ofl5Dh2-LGyABy4ImRT:b-ktslmr) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bc6effc6767a58aef5a12e1b2d1f9a90b1b62a5/pkg/sql/logictest/testdata/logic_test/udf_update#L382)):*
> ```
> ALTER TABLE table_drop DROP COLUMN c;
> 
> statement error pgcode 2BP01 pq: cannot drop column "b" because function "f_update" depends on it
> ```

Looks like we were already handling the case were column is referenced on the RHS of the assignment, correct? It might be good to make the release note and commit message a little more specific about this.
___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-Ofl65xp0ue4SZaijICf:-Ofl65xqAcMq8zCBSIYT:b-uhkb2z) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bc6effc6767a58aef5a12e1b2d1f9a90b1b62a5/pkg/sql/opt/optbuilder/mutation_builder.go#L80)):*
> ```Go
> 	// explicitTargetColOrds contains the ordinals of the columns in targetColSet
> 	// that were explicitly specified by the user, e.g. in the target column list
> 	// of an INSERT or the SET clause of an UPDATE. This is used to determine
> ```

Was there a reproducible bug in `INSERT` too? Or just `UPDATE`?


<!-- Sent from Reviewable.io -->

ZhouXing19 reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice, thanks!

> @ZhouXing19 reviewed 3 of 3 files at r1, 1 of 1 files at r2, 9 of 9 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-:-OfyZThlD_E7XO303jZD:bhuisuw)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

DrewKimball committed:
sql: turn routine dep settings on by default

This commit flips the two new session settings for fixing+improving
routine column dependency tracking to on by default.

Epic: None

Release note: None

DrewKimball reviewed (commented):
>Unfortunately, I think the safe thing to do is make a new session setting given that `use_improved_routine_dependency_tracking` is already on by default in 25.3.

Sad, but true. Done.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-:-Ofyo7EcCv5-XCY05uCe:b-hmi1ph)* status: <span>:shipit:</span> complete! 2 of 0 LGTMs obtained (waiting on @mgartner)

___
*[`pkg/sql/opt/optbuilder/mutation_builder.go` line 80 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-Ofl65xp0ue4SZaijICf:-OfymbED1vzEj-5F-9YU:b-f2k7o7) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bc6effc6767a58aef5a12e1b2d1f9a90b1b62a5/pkg/sql/opt/optbuilder/mutation_builder.go#L80)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Was there a reproducible bug in `INSERT` too? Or just `UPDATE`?
</blockquote></details>

INSERT already tracked dependencies on columns from `insertColIDs`, so just UPDATE.
___
*[`pkg/sql/logictest/testdata/logic_test/udf_update` line 382 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/158935#-Ofl5Dh2-LGyABy4ImRS:-Ofymlwh6pqBVpqw-jPV:b35wj2g) ([raw file](https://github.com/cockroachdb/cockroach/blob/3bc6effc6767a58aef5a12e1b2d1f9a90b1b62a5/pkg/sql/logictest/testdata/logic_test/udf_update#L382)):*
<details><summary><i>Previously, mgartner (Marcus Gartner) wrote…</i></summary><blockquote>

Looks like we were already handling the case were column is referenced on the RHS of the assignment, correct? It might be good to make the release note and commit message a little more specific about this.
</blockquote></details>

Done. Good point.


<!-- Sent from Reviewable.io -->

DrewKimball commented:
TFTRs!

bors r+
